podman/domain/images_manager.py

```diff
@@ -113,18 +113,34 @@ def get_registry_data(
             collection=self,
         )
 
-    def load(self, data: bytes) -> Generator[Image, None, None]:
+    def load(
```

Comment:
```
>>> from podman import PodmanClient
>>> client = PodmanClient()
>>> client.images.load()
<generator object ImagesManager.load at 0xffff6dde4370>
```

Headaches indeed... I am trying to figure out what is wrong here and why the tests are failing.

Reply:
Thanks! Perhaps just I don't see something, but I don't really understand why the exception isn't thrown.

Reply:
got it.

So, this function does not raise

```python
from typing import Optional, Generator

def function(bytearg: Optional[bytes] = None, string: Optional[str] = None) -> Generator[bytes, None, None] | None:
    # no arguments or two are not allowed. Only one can be passed
    if bytearg is None and string is None:
        print("None provided: only one argument is allowed")
        raise ValueError("At least one argument must be provided.")

    if bytearg is not None and string is not None:
        print("Two provided: only one argument is allowed")
        raise ValueError("Only one argument is allowed.")

    if bytearg is not None:
        print("bytearg: {}".format(bytearg))
        yield

    if string is not None:
        print("string: {}".format(string))
        yield

    print("Debug: bytearg: {}, string: {}".format(bytearg, string))
    yield

if __name__ == "__main__":
    function()
```

while this does! why?

```python
from typing import Optional, Generator

def function(bytearg: Optional[bytes] = None, string: Optional[str] = None) -> Generator[bytes, None, None] | None:
    # no arguments or two are not allowed. Only one can be passed
    if bytearg is None and string is None:
        print("None provided: only one argument is allowed")
        raise ValueError("At least one argument must be provided.")

    if bytearg is not None and string is not None:
        print("Two provided: only one argument is allowed")
        raise ValueError("Only one argument is allowed.")

    if bytearg is not None:
        print("bytearg: {}".format(bytearg))
        return

    if string is not None:
        print("string: {}".format(string))
        return

    print("Debug: bytearg: {}, string: {}".format(bytearg, string))
    return

if __name__ == "__main__":
    function()
```
because in the first example the function is a python generator so it `yields` the values. now, with the first snippet try again with

```python

if __name__ == "__main__":
    next(function())
```

the function raises now.

---

Quick solution, make the function return and don't make it raise.

Complicated solution. I don't know right now, I like using exceptions but I am not sure if it's really possible with generators. I'll keep scratching my head on this for a bit.

What do you think @milanbalazs ?

Reply:
Ahh... Good to know! Thanks for your investigation. I have updated the UnitTest implementation. If I put a `return` instead of `PodmanError` exception then the result is the same, actually (Nothing happened). Because of the GeneratorType the `return` or `raise` is "activated" only then the iterator starts, so in my opinion a specific exception is better then a `StopIteration` exception which is thrown by `return` statement.

@inknos What do you think about it?

BTW, currently I am fighting with the PyLint in your gate... It's also a tricky checker :)

Reply:
@inknos 

> while this does! why?
> because in the first example the function is a python generator so it yields the values.

If I understand things correctly, it's because you are creating a generator object, but not "executing it". You'd have to start iterating. By replacing `yield` with `return` it stops being a generator and is actually a regular function which *is* being executed.

Reply:
Shell I change the `Generator` (`yield`) to a simple `return` (and return the list of results. Like: `return [self.get(item) for item in body["Names"]]`)? I didn't want to refactor the "base" mechanism of the method but if you say I should change it, I am willing to do it.

Reply:
I have changed the Generator to a list with return statement. It seems the tests are OK, just the Pylint complains, as I mentioned before.

Reply:
> I have changed the Generator to a list with return statement. It seems the tests are OK, just the Pylint complains, as I mentioned before.

Pylint seems broken on other prs too :/ Ignore it for now.

About the iterator. I probably explained myself wrongly, apologies. We cannot change the return type of the function, or we will break the API. 

If this were the first implementation, then this would be the right path, but we need to keep the consistency here. Please, let's keep the tests with `next()` and the generator return type.

Good news though. I did some research and learned new things. Consider this code:

```python
 def function(bytearg: Optional[bytes] = None, string: Optional[str] = None) -> Generator[bytes, None, None] | None:
     def _generator(bytearg: Optional[bytes] = None, string: Optional[str] = None) -> Generator[bytes, None, None]:
         if bytearg is not None:
             print("bytearg: {}".format(bytearg))
             yield

         if string is not None:
             print("string: {}".format(string))
             yield

         print("Debug: bytearg: {}, string: {}".format(bytearg, string))
         yield

     if bytearg is None and string is None:
         print("None provided: only one argument is allowed")
         raise ValueError("At least one argument must be provided.")

     if bytearg is not None and string is not None:
         print("Two provided: only one argument is allowed")
         raise ValueError("Only one argument is allowed.")

     return _generator(bytearg, string)

if __name__ == "__main__":
    # function() # raises ValueError
    print(function(bytearg='a')) # returns a generator
```

consider this example. now we have a function which returns a generator but it's not a generator itself. so, raising exceptions work as expected and the return statement gets the execution interrupted on the right steps.

wdyt? :) @milanbalazs

Reply:
Nice! Thanks, this is quite useful! I have changed the code and tests based on your suggestion. I have tested and it handles the Generator as well as the exceptions as expected. Let me know if I have to change anything. :)

Reply:
No problem, I like to deliver quality work even if it is a lot of work. :)